core/src/subgraph/instance_manager.rs

```diff
@@ -639,221 +617,195 @@ where
+            &mut ctx,
+            &mut block_state.entity_cache,
+            data_sources,
+            block_ptr_for_new_data_sources,
+        );
 
-        let section = ctx
-            .host_metrics
-            .stopwatch
-            .start_section("entity_cache_evict");
-        cache.evict(*ENTITY_CACHE_SIZE);
-        section.end();
+        // Process the triggers in each host in the same order the
+        // corresponding data sources have been created.
+        for trigger in triggers.into_iter() {
+            block_state = SubgraphInstance::<T>::process_trigger_in_runtime_hosts(
+                &logger,
+                &runtime_hosts,
+                &light_block,
+                trigger,
+                block_state,
+            )
+            .await?;
+        }
+    }
 
-        // Put the cache back in the ctx, asserting that the placeholder cache was not used.
-        assert!(ctx.state.entity_lfu_cache.is_empty());
-        ctx.state.entity_lfu_cache = cache;
+    // Apply entity operations and advance the stream
 
-        if !mods.is_empty() {
-            info!(logger1, "Applying {} entity operation(s)", mods.len());
-        }
+    // Avoid writing to store if block stream has been canceled
+    if block_stream_cancel_handle.is_canceled() {
+        return Err(CancelableError::Cancel);
+    }
 
-        // Transact entity operations into the store and update the
-        // subgraph's block stream pointer
-        let _section = ctx.host_metrics.stopwatch.start_section("transact_block");
-        let subgraph_id = ctx.inputs.deployment_id.clone();
-        let stopwatch = ctx.host_metrics.stopwatch.clone();
-        let start = Instant::now();
-        ctx.inputs
-            .store
-            .transact_block_operations(subgraph_id, block_ptr_after, mods, stopwatch)
-            .map(|should_migrate| {
-                let elapsed = start.elapsed().as_secs_f64();
-                metrics.block_ops_transaction_duration.observe(elapsed);
-                if should_migrate {
-                    ctx.inputs.store.migrate_subgraph_deployment(
-                        &logger1,
-                        &ctx.inputs.deployment_id,
-                        &block_ptr_after,
-                    );
-                }
-                (ctx, needs_restart)
-            })
-            .map_err(|e| {
-                format_err!("Error while processing block stream for a subgraph: {}", e).into()
-            })
-    })
+    let section = ctx.host_metrics.stopwatch.start_section("as_modifications");
+    let ModificationsAndCache {
+        modifications: mods,
+        entity_lfu_cache: mut cache,
+    } = block_state
+        .entity_cache
+        .as_modifications(ctx.inputs.store.as_ref())
+        .map_err(|e| {
+            CancelableError::from(format_err!(
+                "Error while processing block stream for a subgraph: {}",
+                e
+            ))
+        })?;
+    section.end();
```

Comment:
Question for @leoyvens: I see that due to the `?`, we may return early from this. Do we have to clean up the stopwatch section in that case or is it ok to return and never call `section.end()`?

There may be a few more places like that; I just want to double-check that the behavior is fine. We were already returning early before, so I'm assuming this was already ok.

Reply:
It's fine to return early, `section.end()` simply drops `section`, and it will also be dropped on return, ending the section.

Reply:
Ok, cool.